### PR TITLE
Set of API improvements

### LINKED ISSUE #22: Question about default logging on service crash
#### Question : The current default implementation of CriticalBackgroundService.OnError log to the console. Is this a design choice ? 

I'm using your library in one of my windows service application (from a worker service) and a few days ago it crashed. In my log file, I only had "INFO ... Microsoft.Hosting.Lifetime - Application is shutting down...". No reason why. From the lines before the shutdown, I could guess that it crashed during this operation and no log/exception was written. 
I found out that the CriticalBackgroundService.OnError will not log to an ILogger, but to the console. Obviously the console is not redirected to my log file. I'm using Microsoft ILogger, and my ILogger is configured to output to log4net, which log to a file and also the console. 

---
I've implemented the virtual OnError in all my background service to log to my ILogger, so i'm fine with the rest and I don't require any modification from your part. I was kinda not expecting it to log to console, but instead to log to ILogger by default. 
This is probably not because you're targetting a basic program, without any ilogger and all the DI stuff, because you're already registering your services with the microsoft DI in the readme file. 

#### Is it by design that CriticalBackgroundService log to console, while PeriodicTaskRunnerBackgroundService  use the ILogger facility ? 
When I check the rest of your project, PeriodicTaskRunnerBackgroundService seems to use ILogger to report crash, but CriticalBackgroundService only use the console. Perhaps it's because it's critical ?

### LINKED ISSUE #24: Make CriticalBackgroundService.OnError() abstract
### Describe the feature

Users must handle exceptions differently. If they don't need `OnError()`, they don't need `CriticalBackgroundService`.

### LINKED ISSUE #25: Make ensuring TPeriodicTask registered more graceful
### Describe the feature

Call `ServiceProvider.GetService<TPeriodicTask>()` and check result for `null`

### LINKED ISSUE #26: Use factory for instantiating TPeriodicTask
### Describe the feature

Write and use factory instead of direct Service Locator

### LINKED ISSUE #27: Fix CA2254 when logging
### Describe the feature

Move interpolation params from string to `object[]` argument
[Docs](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2254)

### LINKED ISSUE #28: Return IServiceCollection from IServiceCollectionExtensions
### Describe the feature

It will allow to chaining methods.

### PR DESCRIPTION
**Pull request by ProMix0**
Closes #24
Closes #25
Closes #26
Closes #27
Closes #28

Also it unintended resolve #22

### CODE REVIEW

**Comment by ProMix0** [comment]
#30 should help

**Comment by ProMix0** [comment]
This commits will fix build Action

File: src/PeriodicTaskRunnerBackgroundService.cs [hunk 1]
```diff
@@ -48,22 +48,22 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
         {
             // Ensure that we can crate the service. Do this synchronously so that we'll fail-fast no matter the failure mode
             // if the task can't run at all.
-            using (var scope = this.serviceProvider.CreateScope())
+            if(!this.taskFactory.CanResolvePeriodicTask())
             {
-                _ = scope.ServiceProvider.GetRequiredService<TPeriodicTask>();
+                this.logger.LogCritical("{Type} didn't registered in container", typeof(TPeriodicTask));
+                throw new InvalidOperationException($"{typeof(TPeriodicTask)} didn't registered in container");
```
Comment (GeeWee) [inline src/PeriodicTaskRunnerBackgroundService.cs:54]: I see your argument here that the error message isn't very good - I'm not sure why the better way to resolve this issue isn't just to call `scope.ServiceProvider.GetService<TPeriodicTask> != null` in here and then throw the InvalidOperationException?

File: src/PeriodicTaskRunnerBackgroundService.cs [hunk 2]
```diff
@@ -72,12 +72,21 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 
                     if (this.periodicTaskFailureMode == PeriodicTaskFailureMode.RetryLater)
                     {
-                        this.logger.LogWarning(e, $"Exception while processing message in {typeof(TPeriodicTask)}. Retrying in {this.timeBetweenTasks}");
+                        this.logger.LogWarning(e, "Exception while processing message in {Type}. Retrying in {TasksDelay}", typeof(TPeriodicTask), this.timeBetweenTasks);
```
Comment (GeeWee) [inline src/PeriodicTaskRunnerBackgroundService.cs:75]: Good work on the logging templates!

File: src/PeriodicTaskRunnerBackgroundService.cs [hunk 3]
```diff
@@ -48,22 +48,22 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
...
+                throw new InvalidOperationException($"{typeof(TPeriodicTask)} didn't registered in container");
             }
 
             while (!stoppingToken.IsCancellationRequested)
             {
                 try
                 {
-                    using var scope = this.serviceProvider.CreateScope();
-                    var periodicTask = scope.ServiceProvider.GetRequiredService<TPeriodicTask>();
+                    var periodicTask = this.taskFactory.GetPeriodicTask();
```
Comment (GeeWee) [inline src/PeriodicTaskRunnerBackgroundService.cs:61]: I'm interested to hear your thoughts about what the factory does and why it's relevant - I'm not opposed to it, but I'm also not convinced of the merits.

File: src/CriticalBackgroundService.cs [hunk 4]
```diff
@@ -42,12 +42,7 @@ protected CriticalBackgroundService(IApplicationEnder applicationEnder)
         /// but you can override it, if you want to handle it differently, or do some extra logging.
         /// </summary>
         /// <param name="exceptionFromExecuteAsync"></param>
-        protected virtual void OnError(Exception exceptionFromExecuteAsync)
-        {
-            Console.Error.WriteLine($"Error happened while executing CriticalBackgroundTask {this.GetType().FullName}. Shutting down.");
-            Console.Error.WriteLine(exceptionFromExecuteAsync.ToString());
-            this._applicationEnder.ShutDownApplication();
-        }
+        protected abstract void OnError(Exception exceptionFromExecuteAsync);
```
Comment (GeeWee) [inline src/CriticalBackgroundService.cs:45]: I don't think I'm alright with this change - see #24

File: README.md [hunk 5]
```diff
@@ -85,7 +85,7 @@ public class YieldingAndThenCrashingCriticalBackgroundService : CriticalBackgrou
 ## AddHostedServiceAsSingleton
 Hosted Services and BackgroundServices aren't part of the dependency injection container. This means that you can't get them injected into your services or controllers. If you need to do this, you can use the `AddHostedServiceAsSingleton` extension method on the `IServiceCollection`
 
-```
+```csharp
```
Comment (GeeWee) [inline README.md:88]: Nice!

**Review by GeeWee — CHANGES_REQUESTED** [review]
Sorry for taking my sweet time with this.
I think there's a lot of really good stuff here - and thanks for contributing ! I'm not sold on all the changes yet unfortunately, so let's iterate a few times until we get it right!

File: src/PeriodicTaskRunnerBackgroundService.cs (hunk 1 above)
Comment (ProMix0) [reply src/PeriodicTaskRunnerBackgroundService.cs:54]: Sure, it's will also resolve the problem, but I decided to incapsulate requesting services in factory
If I wasn't use factory, I would did as you say

File: src/PeriodicTaskRunnerBackgroundService.cs (hunk 3 above)
Comment (ProMix0) [reply src/PeriodicTaskRunnerBackgroundService.cs:61]: - New functional - user can override standart behavior
- Cleaner code

File: src/CriticalBackgroundService.cs (hunk 4 above)
Comment (ProMix0) [reply src/CriticalBackgroundService.cs:45]: Why?

**Comment by ProMix0** [comment]
I really have no idea why tests (I mean test - `PeriodicTask_ShouldContinueRunningTasks_IfFailureModeIsSetToRetry()`) don't passes

I guess this behavior undefined. But don't know why

**Comment by GeeWee** [comment]
Ah, let me try to dive into the CI situation - that does seem super weird. It might take a week or so before I get around to it though.

**Comment by GeeWee** [comment]
Turns out it's probably going to take longer - real life's getting in the way. You're not forgotten though.

**Comment by ProMix0** [comment]
Maybe I can help?

**Comment by GeeWee** [comment]
I mean absolutely do feel free to dive into the CI issue - there is a chance it's just a flaky test

**Comment by ProMix0** [comment]
Can you try to run CI again? My branch passed tests three times in a row. If now problem isn't resolved... so, I will think a lot :)

**Comment by GeeWee** [comment]
Still seems to be failing..!

**Comment by ProMix0** [comment]
Finally, I found failed test
image
I guess latest commit may fix that issue

File: tests/PeriodicTasksTest.cs [hunk 6]
```diff
@@ -64,21 +65,25 @@ public async Task PeriodicTask_ShouldContinueRunningTasks_IfFailureModeIsSetToRe
...
             // to hope that it's done requesting application shutdown at this point
             for (int i = 0; i < 10; i++)
             {
-                await Task.Delay(100); // 1s ms all in all
+                await Task.Delay(200); // 1s ms all in all
                 await Task.Yield();
             }
 
-            stateHolder.Count.Should().BeGreaterThan(5);
+            stateHolder.Count.Should().BeGreaterThan(2);
```
Comment (GeeWee) [inline tests/PeriodicTasksTest.cs:79]: Did keeping it at 5 make it crash? The task is restarted every 50ms, so with the 2s of delay it should reach 40~ and not just 2

File: tests/PeriodicTasksTest.cs [hunk 7]
```diff
@@ -64,21 +65,25 @@ public async Task PeriodicTask_ShouldContinueRunningTasks_IfFailureModeIsSetToRe
...
-            var client = factory.CreateClient();
+            using var client = factory.CreateClient();
             // Act & assert - we crash here after each invocation, but we truck on. The stateHolder should keep being incremented
 
             // Task is hella flaky because it depends on the internals of the IHostedService - try yielding a bunch of times
             // to hope that it's done requesting application shutdown at this point
             for (int i = 0; i < 10; i++)
             {
-                await Task.Delay(100); // 1s ms all in all
+                await Task.Delay(200); // 1s ms all in all
```
Comment (GeeWee) [inline tests/PeriodicTasksTest.cs:75]:
```suggestion
                await Task.Delay(200); // 2s ms all in all
```

File: tests/CriticalBackgroundServiceTests.cs [hunk 8]
```diff
@@ -42,32 +43,13 @@ public void WhenBackgroundService_ErrorsWithoutYielding_ApplicationCrashes()
...
-                await Task.Yield();
-            }
-
-            // due to https://github.com/dotnet/aspnetcore/issues/25857 we can't test if the process is closed directly
-            applicationEnder.ShutDownRequested.Should().BeTrue();
+            YieldingAndThenCrashingCriticalBackgroundService backgroundService = new(applicationEnder);
 
+            await backgroundService.StartAsync(CancellationToken.None);
 
+            Task.WaitAny(new Task[] { applicationEnder.ShutDownTask }, 3000).Should().Be(0);
```
Comment (GeeWee) [inline tests/CriticalBackgroundServiceTests.cs:52]: Hm, this test doesn't test quite as well the real-life behaviour of the backgorundservice, however it has the advantage of actually passing, so let's roll with that. Good job!

**Review by GeeWee — CHANGES_REQUESTED** [review]
I think I have one spot that could do a little improving, but otherwise we're ready to merge this in! Thanks for sticking with me!

File: tests/CriticalBackgroundServiceTests.cs (hunk 8 above)
Comment (ProMix0) [reply tests/CriticalBackgroundServiceTests.cs:52]: Yes, I know. I think, unit testing is more useful than integration testing in this case

P.S. For example, I use background service without any web context

**Comment by ProMix0** [comment]
I rewrote tests. Yes, it's don't work again
I have a versions that it's a kind of multithread issue

So, I don't know. Tomorrow I will add loggers to tests

**Comment by GeeWee** [comment]
After reflecting on the amount of time I have to maintain this project, I don't think I will have time to handle PR's and stuff unfortunately.

However, if you'd like to make a fork and publish that to nuget under a different name, I would be happy to link to it from this projects README

**Pull request closed without merging**